website/hw/hw4.rst

```diff
@@ -180,103 +218,517 @@ As you are stepping through your program, you might want to be able to
+    Reading /usr/lib/debug//lib/x86_64-linux-gnu/libc-2.31.so from remote target...
+
+And, on the first terminal, the editor will start running.
+
+If you need to quit the editor, we suggest doing so from the first terminal
+(by pressing Ctrl-Q) but you can also force the editor to terminate the
+program by pressing Ctrl-C in GDB, followed by Ctrl-D. Please note that this
+may leave the first terminal in an unstable state (in the same way as described
+above in the case of the editor crashing). If that happens, simply perform
+the steps described earlier.
+
+
+Task 2: Stepping Through the Micro Editor Code
+----------------------------------------------
+
+(40 points)
+
+As you can see in the ``micro/main.c`` file, the ``main()`` function of our
+editor is actually pretty simple and brief (in large part, thanks to the
+modular design of the code which hides the various complexities of
+running an editor). The ``main()`` function basically does the following:
+
+- Enable "raw mode" on the terminal, which will allow the editor to effectively
+  take over the terminal and "draw" characters anywhere on it (when the terminal
+  is running in normal mode, we typically can only print to the terminal line
+  by line)
+- We initialize an ``editor_ctx_t`` struct, which we use as a context object
+  to store global information about the editor.
+- If a file was specified in the command-line when running the ``micro`` executable,
+  we load that file.
+- We set a status message at the bottom of the screen.
+- We enter an "event loop", where we continously repeat the following two operations:
+
+  - We refresh the screen with the most up-to-date content of the editor.
+  - We wait for a keypress (our "event")
+
+  Event loops are a common way of structuring interactive applications, where we just
+  wait for something to happen (in this case, input from the user), and then refresh the
+  contents of the screen based on that event. For example, if the user presses the ``F``
+  key, we would expect an ``F`` to be inserted at the position of the cursor in the editor.
+
+When debugging our code, we may want to see how some of the fields of our context object
+evolve as we use the editor. You must do the following:
+
+1. Open file ``samples/simple.txt`` in the editor. We would like to figure out the
+   exact line(s) in our code where the ``ctx->num_rows``
+   variable is updated when we load a file. On Gradescope, specify the line(s) of code
+   where this happens, and copy-paste the full and unabridged GDB session that helped
+   you determine this.
+2. Open file ``samples/lorem-ipsum.txt`` in the editor. We can press Ctrl-F to search
+   for a piece of text within the file. Internally, the ``editor_find`` function
+   will be called when this happens. Set one or more breakpoints that allow you to determine the
+   values of ``ctx->cx`` and ``ctx->cy`` after searching for the term ``Integer`. On
```

Comment:
For some reason in this task I keep getting this:
```
Warning:
Could not insert hardware watchpoint 3.
Could not insert hardware breakpoints:
You may have requested too many hardware breakpoints/watchpoints.
```

Reply:
What breakpoint did you try to set?

Reply:
[Noting that we've started discussing this on a separate channel]